R/source.R

```diff
@@ -429,33 +434,22 @@ check_cargo_output <- function(compilation_result, message_buffer, tty_has_color
     jsonlite::parse_json
   )
 
-  if (!isTRUE(quiet)) {
-    purrr::walk(
-      gather_cargo_output(
-        cargo_output,
-        "warning",
-        tty_has_colors
-      ),
-      ui_w
-    )
-  }
-
   if (!isTRUE(compilation_result$status == 0)) {
-    error_messages <- purrr::map_chr(
-      gather_cargo_output(
-        cargo_output,
-        "error",
-        tty_has_colors
-      ),
-      bullet_x
+
+    error_messages <- gather_cargo_output(
+      cargo_output,
+      "error",
+      tty_has_colors
     )
 
-    ui_throw(
-      "Rust code could not be compiled successfully. Aborting.",
-      error_messages,
-      call = call,
-      glue_open = "{<{",
-      glue_close = "}>}"
+    cli::cli_abort(
```

Comment:
This is the most difficult part. We want to preserve the ANSI formatting we get from the cargo output so that the user actually sees useful rust compiler hints explaining what is wrong. Just double check that `rust_function()` emits nicely formatted errors if you mess up with your rust code.

Reply:
Honestly, I would say it is not :-( The issue here is the absence of leading spaces, which `{cli}` is so eager to trim. Perhaps we can leave the old implementation just for this particular case? Currently it works like this
```
#> Error in `invoke_cargo()`:
#> ! Rust code could not be compiled successfully. Aborting.
#> x error: expected one of `:`, `@`, or `|`, found `)`
#>  --> src\lib.rs:3:32
#>   |
#> 3 | fn hello_ilia(nowaythiswillwork) {}
#>   |                                ^ expected one of `:`, `@`, or `|`
#>   |
#>   = note: anonymous parameters are removed in the 2018 edition (see RFC 1685)
#> help: if this is a `self` type, give it a parameter name
#>   |
#> 3 | fn hello_ilia(self: nowaythiswillwork) {}
#>   |               +++++
#> help: if this is a parameter name, give it a type
#>   |
#> 3 | fn hello_ilia(nowaythiswillwork: TypeName) {}
#>   |                                ++++++++++
#> help: if this is a type, explicitly ignore the parameter name
#>   |
#> 3 | fn hello_ilia(_: nowaythiswillwork) {}
#>   |               ++
#> x error: expected `:`
#>  --> src\lib.rs:3:32
#>   |
#> 3 | fn hello_ilia(nowaythiswillwork) {}
#>   |                                ^
#> x error[E0425]: cannot find function `meta__hello_ilia` in this scope
#>  --> src\lib.rs:6:4
#>   |
#> 6 | fn hello_ilia;
#>   |    ^^^^^^^^^^ not found in this scope
#> x error: aborting due to 3 previous errors
```
Or directly in the console with ANSI formatting preserved 
![image](https://user-images.githubusercontent.com/8782986/232508572-6f914c31-acf6-41f1-98d4-ac96d7dc76d0.png)

Note how much more useful it is for debugging.

Reply:
Ohhhh!! I see that now. Yeesh! Okay, I'll give this another look.

Reply:
Check if `rlang::last_trace()` looks nice. If it is, then it is a win !